### PR TITLE
Migrate terraform-aws-lb to Terraform 0.12

### PR DESCRIPTION
**Pull request by JonRoma**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+.PHONY: test clean
+
+test: .terraform
+	AWS_DEFAULT_REGION=us-east-2 terraform validate
+	terraform fmt -check
+	! grep -l TF-UPGRADE-TODO *.tf
+	# Check for badge in README.md
+	grep -q "\[\!\[Build Status\]([^)]*)\]([^)]*)" README.md
```
Comment (ddriddle) [inline Makefile]: This Makefile is old. Please update it!

Comment (ddriddle) [reply Makefile]: Also add the Dockerfile.

File: data.tf [hunk 2]
```diff
@@ -1,24 +1,26 @@
-data "aws_region" "current" {}
+data "aws_region" "current" {
```
Comment (ddriddle) [inline data.tf]: Please make this a one-liner.

File: data.tf [hunk 3]
```diff
@@ -1,24 +1,26 @@
...
+}
 
-data "aws_caller_identity" "current" {}
+data "aws_caller_identity" "current" {
```
Comment (ddriddle) [inline data.tf]: Please make this a one-liner.

File: listener.tf [hunk 4]
```diff
@@ -1,17 +1,21 @@
...
-  port              = "${lookup(local.ports[count.index], "port")}"
-  protocol          = "${lookup(local.ports[count.index], "protocol")}"
-  ssl_policy        = "${lookup(local.ports[count.index], "protocol") == "HTTPS" ? lookup(local.ports[count.index], "ssl_policy", var.ssl_policy) : ""}"
-  certificate_arn   = "${lookup(local.ports[count.index], "protocol") == "HTTPS" ? lookup(local.ports[count.index], "certificate_arn", var.certificate_arn) : ""}"
+  count             = length(local.ports)
+  load_balancer_arn = element(concat(aws_lb.default.*.arn, aws_lb.user.*.arn), 0)
+  port              = local.ports[count.index]["port"]
+  protocol          = local.ports[count.index]["protocol"]
+  ssl_policy        = local.ports[count.index]["protocol"] == "HTTPS" ? lookup(local.ports[count.index], "ssl_policy", var.ssl_policy) : ""
+  certificate_arn = local.ports[count.index]["protocol"] == "HTTPS" ? lookup(
```
Comment (ddriddle) [inline listener.tf]: Please make this a one-liner.

File: main.tf [hunk 5]
```diff
@@ -1,37 +1,44 @@
...
-  idle_timeout    = "${var.idle_timeout}"
+  count           = length(keys(var.access_logs)) != 0 ? 1 : 0
+  name            = var.name
+  internal        = var.internal
+  security_groups = concat(var.security_groups, [aws_security_group.default.id])
+  subnets         = data.aws_subnet_ids.selected.ids
+  idle_timeout    = var.idle_timeout
 
-  access_logs = ["${var.access_logs}"]
+  dynamic "access_logs" {
```
Comment (ddriddle) [inline main.tf]: A dynamic block seems like overkill here. Will this work:

```
access_logs = var.access_logs
```

File: outputs.tf [hunk 6]
```diff
@@ -1,25 +1,28 @@
...
 output "arn" {
   description = "The ARN of the load balancer (matches id)."
-  value       = "${element(concat(aws_lb.default.*.arn, aws_lb.user.*.arn), 0)}"
+  value       = element(concat(aws_lb.default.*.arn, aws_lb.user.*.arn), 0)
 }
 
 output "arn_suffix" {
   description = "The ARN suffix for use with CloudWatch Metrics."
-  value       = "${element(concat(aws_lb.default.*.arn_suffix, aws_lb.user.*.arn_suffix), 0)}"
+  value = element(
```
Comment (ddriddle) [inline outputs.tf]: Make this a one-liner please.

File: sg.tf [hunk 7]
```diff
@@ -5,60 +5,74 @@
 #
 resource "aws_security_group" "default" {
   description = "security group for ${var.name} ALB"
-  name        = "${var.name}"
-  vpc_id      = "${data.aws_vpc.selected.id}"
+  name        = var.name
+  vpc_id      = data.aws_vpc.selected.id
 
-  tags = "${merge(map("Name", var.name), var.tags)}"
+  tags = merge(
```
Comment (ddriddle) [inline sg.tf]: Make this a one-liner please.

File: sg.tf [hunk 8]
```diff
@@ -5,60 +5,74 @@
...
+  )
 }
 
 resource "aws_security_group" "secure" {
-  count       = "${length(var.secure_ports)}"
-  description = "${lookup(var.secure_ports[count.index], "security_group_description", "")}"
-  name        = "${lookup(var.secure_ports[count.index], "security_group")}"
-  vpc_id      = "${data.aws_vpc.selected.id}"
+  count = length(var.secure_ports)
+  description = lookup(
```
Comment (ddriddle) [inline sg.tf]: Make this a one-liner please.

Comment (ddriddle) [reply sg.tf]: Is a lookup even needed with v0.12? See line 26.

File: sg.tf [hunk 9]
```diff
@@ -5,60 +5,74 @@
...
+    var.secure_ports[count.index],
+    "security_group_description",
+    "",
+  )
+  name   = var.secure_ports[count.index]["security_group"]
+  vpc_id = data.aws_vpc.selected.id
 
-  tags = "${merge(map("Name", var.name), var.tags)}"
+  tags = merge(
```
Comment (ddriddle) [inline sg.tf]: Make this a one-liner please.

File: target_group.tf [hunk 10]
```diff
@@ -8,12 +8,17 @@
...
+  name = var.name
 
   port     = "80"
   protocol = "HTTP"
 
   target_type = "ip"
-  vpc_id      = "${data.aws_vpc.selected.id}"
-  tags        = "${merge(map("Name", var.name), var.tags)}"
+  vpc_id      = data.aws_vpc.selected.id
+  tags = merge(
```
Comment (ddriddle) [inline target_group.tf]: Make this a one-liner please.

File: README.md [hunk 11]
```diff
@@ -8,7 +10,7 @@ Example Usage
 ### Public load balancer
 ```hcl
 module "lb_name" {
-  source = "git@github.com:cites-illinois/as-aws-modules//lb"
+  source = "git@github.com:techservicesillinois/terraform-aws-lb//"
```
Comment (ddriddle) [inline README.md]: Please drop the double slashes '//' at the end of the source line.

File: README.md [hunk 12]
```diff
@@ -44,10 +46,10 @@ module "lb_name" {
 ### Private load balancer
 ```hcl
 module "lb_name" {
-  source = "git@github.com:cites-illinois/as-aws-modules//lb"
+  source = "git@github.com:techservicesillinois/terraform-aws-lb//"
```
Comment (ddriddle) [inline README.md]: Please drop the double slashes '//' at the end of the source line.

File: main.tf (hunk 5 above)
Comment (ddriddle) [reply main.tf]: I recommend you try testing the module by creating a load balancer without access logs.

Comment (JonRoma) [reply main.tf]: As far as I can tell, the behavior of the migrated module is one-for-one with the current behavior, and I have the tests to prove it.

I've filed an issue #6 for future work to streamline the lb code and make it more flexible. However it is likely that any refactoring we do in the future will be a breaking change due to the structural changes to the Terraform state (or at the very least will require using `terraform import`).

File: sg.tf (hunk 8 above)
Comment (JonRoma) [reply sg.tf]: See my comment above about consistency with the old module's behavior, and about issue #6.

**Pull request merged**